Add rounded rectangle shape

msiglreith opened the pull request:
Rectangle shape with rounded corners (all corners same radius).

TODO:
- [x] Winding (not sure what to do here)
- [x] More accurate area estimate (handle invalid states radius > width/2?)
- [x] Path generation

raphlinus commented on src/rounded_rect.rs:
```diff
@@ -0,0 +1,138 @@
+    /// Coordinates of the rectangle.
+    pub rect: Rect,
+    /// Radius of all four corners.
+    pub radius: f64,
```
An interesting question, should there be a single radius value or one for each corner? I think I would lean toward the latter as it's more general and I can easily imagine such things used in UI.

msiglreith replied:
Good point, I currently focused on one radius for all corners as this is natively supported by d2d (actually they use ellipses but not sure how useful that is). Nonetheless, the backend could decide internally which path to use for rendering.

raphlinus commented on src/rounded_rect.rs:
```diff
@@ -0,0 +1,138 @@
+
+    /// Take absolute value of width, height and radius.
+    ///
+    /// The resulting rect has the same extents as the original, but is
```
What happens when the radius has different sign as the area? I'm wondering whether this case should be forbidden by construction.

raphlinus commented on src/rounded_rect.rs:
```diff
@@ -0,0 +1,138 @@
+    }
+
+    #[inline]
+    fn bounding_box(&self) -> Rect {
```
Note that this will require `abs` if we allow negative area rectangles, and concern for out-of-range radius.

msiglreith commented:
Rebased on master and implemented the missing pieces. Please have a look again (:
I moved the arc part into a separate module as I use it for implementing the path generation for the rounded rectangle.

raphlinus commented on src/rounded_rect.rs:
```diff
@@ -0,0 +1,302 @@
+    /// Absolute radius of the rounded corners.
+    ///
+    /// Clamped to half the size of the rectangle.
+    pub fn radius(&self) -> f64 {
```
I think it would be better do this on construction.

raphlinus commented on src/rounded_rect.rs:
```diff
@@ -0,0 +1,302 @@
+    ///
+    /// The result will have non-negative width and height.
+    #[inline]
+    pub fn from_origin_size(origin: Point, size: Vec2, radius: f64) -> RoundedRect {
```
The method in `Rect` that this is copied from is wrong, the `size` argument can have negative components :/

raphlinus commented on src/rounded_rect.rs:
```diff
@@ -0,0 +1,302 @@
+    }
+
+    #[test]
+    fn winding() {
```
I think if this was tested on a rect not centered at the origin, it would reveal more problems.

raphlinus commented on src/rounded_rect.rs:
```diff
@@ -0,0 +1,302 @@
+
+    fn next(&mut self) -> Option<PathEl> {
+        if self.idx > 4 {
+            return None;
```
Does this miss the `ClosePath` at the end? I'm finding it a bit hard to follow the logic.

msiglreith replied:
Yes, also verified it visually. Added a few more inline comments for better explanation.

raphlinus commented on src/rounded_rect.rs:
```diff
@@ -0,0 +1,302 @@
+        let inside_half_height = (self.height() / 2.0 - self.radius()).max(0.0);
+
+        // project point out of the inner rectangle (positive quadrant)
+        let px = (pt.x.abs() - inside_half_width).max(0.0);
```
I think you need to be subtracting the center from `pt` first; then I believe the logic is probably correct, but haven't validated it carefully.

msiglreith commented:
Thanks for the great review!
Fixed the issues and improved documentation a bit.

cmyr commented on src/rounded_rect.rs:
```diff
@@ -0,0 +1,305 @@
+impl RoundedRect {
+    /// A new rectangle from minimum and maximum coordinates.
+    #[inline]
+    pub fn new(x0: f64, y0: f64, x1: f64, y1: f64, radius: f64) -> RoundedRect {
```
Two thoughts:

- I would add a new constructor that takes a `Rect` and a `radius`.
- In `RoundedRect::new`, I would just call `Rect::new`, and then I would call the new constructor mentioned above.

msiglreith replied:
Thanks, done!

msiglreith commented on src/arc.rs:
```diff
@@ -0,0 +1,113 @@
+
+/// A single arc segment.
+#[derive(Clone, Copy, Debug)]
+pub struct Arc {
```
This has been made public recently and also exported as part of the svg module.

msiglreith commented:
ping, any open tasks left for me (:?
